Implement contest settings short display

Ninjaclasher opened the pull request:
This fixes #695 and supercedes #700.

We add an option to turn off this short display so that information is not duplicated for existing contests.

Todo:
 - [x] Add contest format-specific information.

Ninjaclasher commented:
Screenshot: 
![image](https://user-images.githubusercontent.com/29607503/122647439-38e31000-d0f2-11eb-9761-4e6a254c8e81.png)

kiritofeng commented on judge/contest_format/ecoo.py:
```diff
@@ -125,3 +125,25 @@ def display_participation_result(self, participation):
+        time_bonus = self.config['time_bonus']
+        if time_bonus:
+            yield ungettext(
+                'For every **1 minute** you submit before the end of your window, there will be a **1** point bonus.',
```
Why do you hardcode the `1` here but no where else?

Ninjaclasher replied:
fixed

kiritofeng commented on judge/views/contests.py:
```diff
@@ -234,8 +234,23 @@ def get_context_data(self, **kwargs):
+            **self.object.contest_problems
+            .annotate(
+                partials_enabled=F('partial').bitand(F('problem__partial')),
+                pretests_enabled=F('is_pretested').bitand(F('contest__run_pretests_only')),
```
Can someone confirm that this will work even after systesting? cc @Xyene @quantum5

Ninjaclasher replied:
This unfortunately won't be correct after systests, but I don't see any way to correctly grab that information after the settings are unchecked. FWIW this is also how it was done in #700, so it might be fine (?)

quantum5 replied:
This is a defect in the design of the pretest system. What we really want is two different booleans, but that's out of scope here.

kiritofeng commented on templates/contest/contest.html:
```diff
@@ -65,6 +65,115 @@
+                        {% endif %}
+                    {% endwith %}
+                    <li>
+                        {% if contest.is_rated %}
```
Is there a way to clean this up, say in the view using formatted strings?

Ninjaclasher replied:
We could break up different segments of that sentence and join them together, but I'm not sure if that will make translations possible/correct. Alternatively, we could use a nested list and list the rating ceiling/floor restrictions there, but I originally opted to not go that route as it may be confusing to a user to see "rated for all participants" followed by some rating restrictions.

kiritofeng replied:
i18n is pain...

this gets an "ok fine then" from me, I'll see what others have to say on this

quantum5 replied:
We definitely must use discrete sentences or colon-style for i18n. Colon-style will be confusing, so I think this is the best we can do.

kiritofeng commented on templates/contest/contest.html:
```diff
@@ -65,6 +65,115 @@
+                    <li>
+                        {% filter markdown('default') %}
+                            {% trans trimmed count=metadata.problem_count %}
+                                There is **1** problem in this contest.
```
Is hardcoded 1 a good idea? (i18n is pain)

Ninjaclasher replied:
fixed

kiritofeng reviewed (changes requested):
Potential i18n problems are pain~~

maxcruickshanks reviewed (changes requested):
There is a typo for penalties: ```Each submission before the first maximum score submission with incur a **penalty of %d minute**.```. It should be ```will incur``` instead of ```with incur```.

Markdown is not capitalized: ```Returns a generator of markdown strings```.

A is not supposed to be capitalized: ```:return A generator, where each item is an individual line.```

```The scoreboard will be **hidden** for the duration of the contest.``` seems vague.
Roger's potential replacement: ```the entire duration of the contest```.

Ninjaclasher commented:
> There is a typo for penalties: `Each submission before the first maximum score submission with incur a **penalty of %d minute**.`. It should be `will incur` instead of `with incur`.

Fixed.

> Markdown is not capitalized: `Returns a generator of markdown strings`.

Fixed. 

> A is not supposed to be capitalized: `:return A generator, where each item is an individual line.`

The `A` is supposed to be captialized (see the other docstrings). However, I was missing a `:` after `:return`, which is now added.

> `The scoreboard will be **hidden** for the duration of the contest.` seems vague.
> Roger's potential replacement: `the entire duration of the contest`.

Updated

kiritofeng reviewed (approved):
LGTM

quantum5 reviewed (approved):
Looks good, will deploy after current rated contest finishes.

Ninjaclasher commented:
Rebased migration.

Ninjaclasher commented:
Rebased migration again.
